keps/672-execution-policy/README.md

```diff
@@ -0,0 +1,389 @@
+```
+
+#### Story 3
+
+As an HPC user, I want to run a set of simulations with MPI that complete successfully
+(or some percentage) before using the data for a next step analysis.
+
+As an HPC user I want to scale up my simulation job only after a small set have completed
+successfully. For that case, I could imagine essentially the same replicatedJob done twice,
+just with a larger size. If the outcome of the first isn't success you wouldn't launch
+the larger bulk of work.
+
+### Risks and Mitigations
+
+This API will not allow to describe DAGs to avoid workflow manager features in JobSet.
+The goal is to only focus on Job sequence to cover model training/HPC use-cases.
+
+## Design Details
+
+### API Details
+
+```golang
+type JobSetSpec struct {
+	ExecutionPolicy *ExecutionPolicy `json:"executionPolicy,omitempty"`
+}
+
+type ExecutionPolicyOption string
+
+const (
+	// This is the default settings.
+	// AnyOrder means that Jobs will be started in any order.
+	AnyOrder ExecutionPolicyOption = "AnyOrder"
+
+	// InOrder starts the ReplicatedJobs in order that they are listed. Jobs within a ReplicatedJob
+	// will still start in any order.
+	InOrder ExecutionPolicyOption = "InOrder"
+)
+
+type ExecutionPolicy struct {
+	// Order in which Jobs will be created.
+	ExecutionPolicyOrder ExecutionPolicyOption `json:"executionPolicyOrder"`
+
+	// After all ReplicatedJobs reach this status, the JobSet will create the next ReplicatedJobs.
+	ExecutionPolicyRule []ExecutionPolicyRule `json:"rules"`
+}
+
+// ExecutionPolicyRule represents the execution policy rule for Job sequence.
+type ExecutionPolicyRule struct {
+
+	// Names of the replicated Jobs that applied the status.
+	TargetReplicatedJobs []string `json:"targetReplicatedJobs"`
+
+	// Status the target ReplicatedJobs must reach before subsequent ReplicatedJobs begin executing.
+	ReplicatedJobsStatus ReplicatedJobsStatusOption `json:"replicatedJobsStatus"`
+}
+
+type ReplicatedJobsStatusOption string
+
+// TODO: What statuses do we want to support in the first version ?
+const (
+	ReadyStatus ReplicatedJobsStatusOption = "Ready"
+
+	SucceededStatus ReplicatedJobsStatusOption = "Succeeded"
+
+	FailedStatus ReplicatedJobsStatusOption = "Failed"
+)
+```
+
+### Implementation
+
+Open Questions:
+
+1. How we should calculate quota for JobSet with execution policy ?
+2. Integration with Kueue. Related KEP to support Argo Workflow in Kueue: https://github.com/kubernetes-sigs/kueue/pull/2976
+
+The JobSet operator will control the creation of ReplicatedJobs based on their status. When
+desired replicated Jobs reach the status defined in replicatedJobsStatus, the controller creates
+the next set of replicated Jobs.
+
+If JobSet is suspended the all ReplicatedJobs will be suspended and the Job sequence starts again.
+
+### Defaulting/Validation
+
+- ExecutionPolicy is immutable.
+- ExecutionPolicyOrder of `AnyOrder` is default setting.
+- StartupPolicy should be empty if ExecutionPolicy is set
```

Comment:
I think StartupPolicy uses AnyOrder actually. So you would need to validate that StartupPolicy is not InOrder.

Reply:
Good point.